martin-mbtiles/src/mbtiles.rs

```diff
@@ -378,33 +388,105 @@ impl Mbtiles {
         Err(MbtError::NoUniquenessConstraint(self.filepath.clone()))
     }
 
-    pub async fn validate_mbtiles<T>(&self, conn: &mut T) -> MbtResult<()>
+    async fn get_global_hash<T>(&self, conn: &mut T) -> MbtResult<Option<String>>
     where
         for<'e> &'e mut T: SqliteExecutor<'e>,
     {
+        let rusqlite_conn = RusqliteConnection::open(Path::new(&self.filepath()))?;
+        register_md5_function(&rusqlite_conn)?;
         let mbttype = self.detect_type(&mut *conn).await?;
 
-        let sql = match mbttype {
+        Ok(match mbttype {
             MbtType::Flat => {
-                return Err(IncorrectDataFormat(
-                    self.filepath().to_string(),
-                    &[MbtType::FlatWithHash, MbtType::Normalized],
-                    MbtType::Flat,
-                ));
-            }
-            MbtType::FlatWithHash => {
-                "SELECT * FROM tiles_with_hash WHERE tile_hash!=hex(md5(tile_data)) LIMIT 1;"
+                println!("Cannot generate global hash, no hash column in flat table format. Skipping global_hash generation...");
+                None
             }
-            MbtType::Normalized => {
-                "SELECT * FROM images WHERE tile_id!=hex(md5(tile_data)) LIMIT 1;"
+            MbtType::FlatWithHash => Some(rusqlite_conn.query_row_and_then("SELECT hex(md5_concat(cast(zoom_level AS text), cast(tile_column AS text), cast(tile_row AS text), tile_hash)) FROM tiles_with_hash ORDER BY zoom_level, tile_column, tile_row;",[], |row| row.get::<_, String>(0))?),
+            MbtType::Normalized => Some(rusqlite_conn.query_row_and_then("SELECT hex(md5_concat(cast(zoom_level AS text), cast(tile_column AS text), cast(tile_row AS text), tile_id)) FROM map ORDER BY zoom_level, tile_column, tile_row;",[], |row| row.get::<_, String>(0))?)
+        })
+    }
+
+    pub async fn generate_global_hash<T>(&self, conn: &mut T) -> MbtResult<()>
+    where
+        for<'e> &'e mut T: SqliteExecutor<'e>,
+    {
+        if let Some(global_hash) = self.get_global_hash(&mut *conn).await? {
+            query!(
+                "INSERT OR REPLACE INTO metadata(name, value) VALUES('global_hash', ?)",
+                global_hash
+            )
+            .execute(conn)
+            .await?;
+        }
+
+        Ok(())
```

Comment:
you don't need that new sqlx file if you do this instead - which is also better because it doesn't introduce duplicated code for the same functionality:

```suggestion
        if let Some(global_hash) = self.get_global_hash(&mut *conn).await? {
            self.set_metadata_value(conn, "global_hash", Some(global_hash))
                .await
        } else {
            Ok(())
        }
```